src/rgen/Microsoft.Macios.Generator/DataModel/BindingInfo.cs

```diff
@@ -18,42 +18,60 @@ namespace Microsoft.Macios.Generator.DataModel;
 
 	public BindingType BindingType => bindingType;
 
-	public BindingData (BindingType type, BindingTypeData data)
+	public BindingInfo (BindingType type, BindingTypeData data)
 	{
 		bindingType = type;
 		bindingTypeData = data;
 	}
 
-	public BindingData (BindingTypeData<ObjCBindings.Class> data)
+	public BindingInfo (BindingTypeData<ObjCBindings.Class> data)
 	{
 		bindingType = BindingType.Class;
 		classData = data;
 	}
 
-	public BindingData (BindingTypeData<ObjCBindings.Protocol> data)
+	public BindingInfo (BindingTypeData<ObjCBindings.Protocol> data)
 	{
 		bindingType = BindingType.Protocol;
 		protocolData = data;
 	}
 
-	public BindingData (BindingTypeData<ObjCBindings.Category> data)
+	public BindingInfo (BindingTypeData<ObjCBindings.Category> data)
 	{
 		bindingType = BindingType.Category;
 		categoryData = data;
 	}
 
-	public static implicit operator BindingTypeData (BindingData data) => data.bindingTypeData;
-	public static implicit operator BindingTypeData<ObjCBindings.Class> (BindingData data) => data.classData;
-	public static implicit operator BindingTypeData<ObjCBindings.Protocol> (BindingData data) => data.protocolData;
-	public static implicit operator BindingTypeData<ObjCBindings.Category> (BindingData data) => data.categoryData;
+	public static implicit operator BindingTypeData (BindingInfo info) => info.bindingTypeData;
 
-	public static implicit operator BindingData (BindingTypeData data) => new (BindingType.Unknown, data);
-	public static implicit operator BindingData (BindingTypeData<ObjCBindings.Class> data) => new (data);
-	public static implicit operator BindingData (BindingTypeData<ObjCBindings.Protocol> data) => new (data);
-	public static implicit operator BindingData (BindingTypeData<ObjCBindings.Category> data) => new (data);
+	public static implicit operator BindingTypeData<ObjCBindings.Class> (BindingInfo info)
+	{
+		if (info.BindingType != BindingType.Class)
+			throw new InvalidCastException ($"Invalid cast to ObjCBindings.Class for binding type {info.BindingType}");
+		return info.classData;
+	}
+
+	public static implicit operator BindingTypeData<ObjCBindings.Protocol> (BindingInfo info)
+	{
+		if (info.BindingType != BindingType.Protocol)
+			throw new InvalidCastException ($"Invalid cast to ObjCBindings.Protocol for binding type {info.BindingType}");
+		return info.protocolData;
+	}
+
+	public static implicit operator BindingTypeData<ObjCBindings.Category> (BindingInfo info)
+	{
+		if (info.BindingType != BindingType.Category)
+			throw new InvalidCastException ($"Invalid cast to ObjCBindings.Category for binding type {info.BindingType}");
```

Comment:
There are three invalid cast exceptions that all fall the same pattern: `$"Invalid cast to {targetType} for binding type {info.BindingType}"`
Suggest:
```
InvalidCastException BindingCastException(string targetType, BindingInfo info)
{
    return new InvalidCastException("$Invalid cast to {targetType} for binding type {info.BindingType}");
}
```

And if you don't care about static analysis, you could also do:
```
void ThrowOnInvalidBindingType(BindingType bindingType, BindingInfo info, string targetType)
{
    if (info.BindingType != bindingType) {
        throw new InvalidCastException("$Invalid cast to {targetType} for binding type {info.BindingType}");
    }
}
```